Description: Use tree timestamp when exporting.
 .
 This helps generate consistent hashes.
Origin: commit, revision id: jelmer@jelmer.uk-20190113173222-dteocq7209l6y0m6
Author: Jelmer VernooÄ³ <jelmer@jelmer.uk>
Bug: https://launchpad.net/bugs/262261
Last-Update: 2019-01-13
Applied-Upstream: no
X-Bzr-Revision-Id: jelmer@jelmer.uk-20190113173222-dteocq7209l6y0m6

=== modified file 'breezy/archive/tar.py'
--- old/breezy/archive/tar.py	2018-11-16 18:33:17 +0000
+++ new/breezy/archive/tar.py	2019-01-13 17:32:22 +0000
@@ -137,8 +137,7 @@
         buf = BytesIO()
         zipstream = gzip.GzipFile(basename, 'w', fileobj=buf,
                                   mtime=root_mtime)
-        for chunk in tarball_generator(
-                tree, root, subdir, force_mtime):
+        for chunk in tarball_generator(tree, root, subdir, force_mtime):
             zipstream.write(chunk)
             # Yield the data that was written so far, rinse, repeat.
             yield buf.getvalue()

=== modified file 'breezy/export.py'
--- old/breezy/export.py	2018-11-16 18:33:17 +0000
+++ new/breezy/export.py	2019-01-13 17:32:22 +0000
@@ -64,7 +64,11 @@
         root = get_root_name(dest)
 
     if not per_file_timestamps:
-        force_mtime = time.time()
+        if getattr(tree, '_repository', None):
+            force_mtime = tree._repository.get_revision(
+                tree.get_revision_id()).timestamp
+        else:
+            force_mtime = time.time()
     else:
         force_mtime = None
 

=== modified file 'breezy/tests/test_export.py'
--- old/breezy/tests/test_export.py	2018-11-11 04:08:32 +0000
+++ new/breezy/tests/test_export.py	2019-01-13 17:32:22 +0000
@@ -16,6 +16,7 @@
 
 """Tests for breezy.export."""
 
+import gzip
 import os
 import tarfile
 import time
@@ -215,6 +216,18 @@
         tf = tarfile.open('target.tar.gz')
         self.assertEqual(["target/a"], tf.getnames())
 
+    def test_tgz_consistent_mtime(self):
+        wt = self.make_branch_and_tree('.')
+        self.build_tree(['a'])
+        wt.add(["a"])
+        timestamp = 1547400500
+        revid = wt.commit("1", timestamp=timestamp)
+        revtree = wt.branch.repository.revision_tree(revid)
+        export.export(revtree, 'target.tar.gz', format="tgz")
+        with gzip.GzipFile('target.tar.gz', 'r') as f:
+            f.read()
+            self.assertEqual(int(f.mtime), timestamp)
+
     def test_tgz_ignores_dest_path(self):
         # The target path should not be a part of the target file.
         # (bug #102234)

